gadget/install/content.go

```diff
@@ -120,22 +121,65 @@ func writeFilesystemContent(laidOut *gadget.LaidOutStructure, kSnapInfo *KernelS
 	if laidOut.Role() == gadget.SystemData && kSnapInfo != nil && kSnapInfo.NeedsDriversTree {
 		destRoot := mountpoint
 		if kSnapInfo.IsCore {
-			destRoot = filepath.Join(mountpoint, "system-data")
+			// For core we write the changes in _writable_defaults. The
+			// files are then copied from the initramfs by
+			// populate-writable.service on first boot as the directories
+			// they are in are marked as "transitional" in the
+			// writable-paths file. We cannot copy directly to
+			// "system-data" as that would prevent files already in
+			// _writable_defaults in the directories of interest to not
+			// be copied (because the files are not copied if the
+			// directory exists already).
+			destRoot = filepath.Join(mountpoint, "system-data", "_writable_defaults")
 		}
 		destDir := kernel.DriversTreeDir(destRoot, kSnapInfo.Name, kSnapInfo.Revision)
 		logger.Noticef("building drivers tree in %s", destDir)
 
-		kTargetSysMntPt := snap.MinimalSnapContainerPlaceInfo(kSnapInfo.Name, kSnapInfo.Revision)
+		cpi := snap.MinimalSnapContainerPlaceInfo(kSnapInfo.Name, kSnapInfo.Revision)
 		if err := kernelEnsureKernelDriversTree(
 			kernel.MountPoints{
 				Current: kSnapInfo.MountPoint,
-				Target:  kTargetSysMntPt.MountDir(),
+				Target:  cpi.MountDir(),
 			},
 			nil,
 			destDir,
 			&kernel.KernelDriversTreeOptions{KernelInstall: true}); err != nil {
 			return err
 		}
+
+		// Create mount unit to make the kernel snap content available from
+		// the drivers tree.
+		squashfsPath := dirs.StripRootDir(cpi.MountFile())
+		whereDir := dirs.StripRootDir(cpi.MountDir())
+
+		hostFsType, options := systemd.HostFsTypeAndMountOptions("squashfs")
+		mountOptions := &systemd.MountUnitOptions{
+			Lifetime:                 systemd.Persistent,
+			Description:              cpi.MountDescription(),
+			What:                     squashfsPath,
+			Where:                    whereDir,
+			Fstype:                   hostFsType,
+			Options:                  options,
+			MountUnitType:            systemd.BeforeDriversLoadMountUnit,
+			RootDir:                  destRoot,
+			PreventRestartIfModified: true,
+		}
+		unitFileName, _, err := systemd.EnsureMountUnitFileContent(mountOptions)
```

Comment:
will this unit be replaced by the one created during seeding ? what happens in a preseed scenario instead?

Reply:
Yes, this unit is the same as the one created while seeding. Should be created in the preseed scenario as well.

Reply:
I don't remember in the preseed case in which order we consume _writable_defaults or whether we don't consume it all ?

Reply:
I don't think we need to consume it on preseeding. My point is that the things we are creating here (drivers tree and mount units for the kernel snap) are created by preseeding as preseeding creates these files on snap "installation". Writable paths are taken into account from initramfs on first boot, what will happen is that if there is a preseeding tarball, the files from the preseeding tarball will be overwritten by the files generated here, which should be fine as content should be identical. Maybe we could find out if we have a preseeding tarball and do not proceed to creating these files, I've added a comment about this. Although in most cases we would install from initramfs if we have a preseeding tarball.